### PR TITLE
Implement additional i8x16 shift instructions

### PR DESCRIPTION
**Pull request by abrown**
This is a follow-on to #1377; review once that PR is merged. This PR adds support for i8x16 left shift and arithmetic right shift. Both are long sequences of instructions on x86.

### CODE REVIEW

**Comment by abrown** [comment]
Ok, #1377 is merged; asked @julian-seward1 for a review.

File: cranelift/codegen/src/isa/x86/enc_tables.rs [hunk 1]
```diff
@@ -1385,6 +1385,73 @@ fn convert_ushr(
...
+        if arg0_type == I8X16 {
+            // First, shift the vector using an I16X8 shift.
+            let bitcasted = pos.ins().raw_bitcast(I16X8, arg0);
+            let shifted = pos.ins().x86_psll(bitcasted, shift_index);
+            let shifted = pos.ins().raw_bitcast(I8X16, shifted);
+
+            // Then, fixup the even lanes that have incorrect lower bits. This uses the 128 mask
+            // bytes as a table that we index into. It is a substantial code-size increase but
+            // reduces the instruction count slightly.
+            let masks = pos.func.dfg.constants.insert(SHL_MASKS.as_ref().into());
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x86/enc_tables.rs:1439]: Read naively, this would seem to insert the mask table once for every converted ishl.  That would be ungood.  Does `pos.func.dfg.constants.insert` de-dup, and hence keep just one copy?

File: cranelift/codegen/meta/src/isa/x86/instructions.rs [hunk 2]
```diff
@@ -376,6 +376,40 @@ pub(crate) fn define(
...
+    let y = &Operand::new("y", TxN);
+    let a = &Operand::new("a", TxN);
+
+    ig.push(
+        Inst::new(
+            "x86_punpckh",
+            r#"
+        Unpack the high-order lanes of ``x`` and ``y`` and interleave into ``a``. With notional
+        i8x4 vectors, where ``x = [x1, x2, x3, x4]`` and ``y = [y1, y2, y3, y4]``, this operation
+        would result in ``a = [x3, y3, x4, y4]`` (note that this uses left-to-right lane ordering
```
Comment (julian-seward1) [inline cranelift/codegen/meta/src/isa/x86/instructions.rs:388]: What does `this` refer to here?  On the whole I'd prefer to stick with the Intel style convention, where the rightmost lane is lane zero.  But note; my comment is very "drive-by"; I don't know what the lane numbering conventions in CL are.

File: cranelift/codegen/meta/src/isa/x86/legalize.rs [hunk 3]
```diff
@@ -484,13 +493,37 @@ fn define_simd(shared: &mut SharedDefinitions, x86_instructions: &InstructionGro
...
+    }
+    {
+        let sshr = sshr.bind(vector(I8, sse_vector_size));
+        let bitcast_i64x2 = bitcast.bind(vector(I64, sse_vector_size));
+        let raw_bitcast_i16x8 = raw_bitcast.bind(vector(I16, sse_vector_size));
+        let raw_bitcast_i16x8_again = raw_bitcast.bind(vector(I16, sse_vector_size));
+        narrow.legalize(
+            def!(z = sshr(x, y)),
+            vec![
+                // Since we will use the high byte of each 16x8 lane, shift an extra 16 bits.
```
Comment (julian-seward1) [inline cranelift/codegen/meta/src/isa/x86/legalize.rs:513]: Did you mean "shift an extra 8 bits" ?

File: cranelift/codegen/src/isa/x86/enc_tables.rs (hunk 1 above)
Comment (abrown) [reply cranelift/codegen/src/isa/x86/enc_tables.rs:1439]: `ConstantPool` [does de-duplicate](https://github.com/bytecodealliance/wasmtime/blob/a148de08f8b4ccdb6e79410cb4b3169b44328201/cranelift/codegen/src/ir/constant.rs#L220-L231) values but currently it is limited to a per-function boundary because of how things are compiled. I opened #1385 to talk about this some more: I think with a bit of thought we could de-duplicate across modules at least.

File: cranelift/codegen/meta/src/isa/x86/instructions.rs (hunk 2 above)
Comment (abrown) [reply cranelift/codegen/meta/src/isa/x86/instructions.rs:388]: Ok, we can switch to Intel-style. I went back and forth on this one and I can't recall why I chose this ordering. I'll look around the file to see if there was something that made me choose this.

File: cranelift/codegen/meta/src/isa/x86/legalize.rs (hunk 3 above)
Comment (abrown) [reply cranelift/codegen/meta/src/isa/x86/legalize.rs:513]: Yup, good catch.
